### PR TITLE
Promlint fixes

### PR DESCRIPTION
**Pull request by abhide**
Updates #2554.

### CODE REVIEW

File: internal/metrics/metrics.go [hunk 1]
```diff
@@ -58,11 +64,17 @@ type Meta struct {
...
-	HTTPProxyInvalidGauge   = "contour_httpproxy_invalid_total"
-	HTTPProxyValidGauge     = "contour_httpproxy_valid_total"
-	HTTPProxyOrphanedGauge  = "contour_httpproxy_orphaned_total"
+	DeprecatedHTTPProxyTotalGauge     = "contour_httpproxy_total"
+	DeprecatedHTTPProxyRootTotalGauge = "contour_httpproxy_root_total"
+	DeprecatedHTTPProxyInvalidGauge   = "contour_httpproxy_invalid_total"
+	DeprecatedHTTPProxyValidGauge     = "contour_httpproxy_valid_total"
+	DeprecatedHTTPProxyOrphanedGauge  = "contour_httpproxy_orphaned_total"
+
+	HTTPProxyTotalGauge     = "contour_httpproxy"
```
Comment (sunjayBhatia) [inline internal/metrics/metrics.go:73]: Could you add some tests to cover that the deprecated and new metric names are the same value? We should add this coverage to ensure we don't break anyone who is already using these metrics before we remove them

File: go.mod [hunk 2]
```diff
@@ -4,18 +4,15 @@ go 1.15
 
 require (
 	github.com/ahmetb/gen-crd-api-reference-docs v0.2.0
-	github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751 // indirect
-	github.com/alecthomas/units v0.0.0-20190717042225-c3de453c63f4 // indirect
 	github.com/envoyproxy/go-control-plane v0.9.8
-	github.com/golang/protobuf v1.4.2
+	github.com/golang/protobuf v1.4.3
```
Comment (sunjayBhatia) [inline go.mod:8]: I forget what our policy on go dependency bumps is but we might want to do this in a separate PR? cc @projectcontour/maintainers

File: hack/generate-metrics-doc.go [hunk 3]
```diff
@@ -78,4 +94,6 @@ func main() {
 
 		f.Close()
 	}
+
+	runPromlint(family)
```
Comment (sunjayBhatia) [inline hack/generate-metrics-doc.go:98]: Maybe comment this out for now with a TODO (and link a relevant issue) since this will cause the code generation job to fail

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]

File: go.mod (hunk 2 above)
Comment (youngnick) [reply go.mod:8]: Generally, we prefer dependency bumps in separate PRs, if this is required for this change though, I think it's okay this time.

Comment (abhide) [reply go.mod:8]: Filed issue: https://github.com/projectcontour/contour/issues/3335
PR: https://github.com/projectcontour/contour/pull/3336

File: hack/generate-metrics-doc.go (hunk 3 above)
Comment (abhide) [reply hack/generate-metrics-doc.go]: fixed.

File: internal/metrics/metrics.go (hunk 1 above)
Comment (abhide) [reply internal/metrics/metrics.go:73]: yes. updated an existing test.

File: go.mod (hunk 2 above)
Comment (abhide) [reply go.mod:8]: Yes, bunch of stuff is getting changed if I update prometheus golang library.

**Review by sunjayBhatia — APPROVED** [review]
LGTM as far as the bits I know about for metrics docs and generated bits + tests

**Comment by skriss** [comment]
@abhide this needs a rebase

**Comment by abhide** [comment]
@skriss rebased and pushed another change.

File: internal/metrics/metrics.go [hunk 4]
```diff
@@ -85,6 +97,41 @@ func NewMetrics(registry *prometheus.Registry) *Metrics {
 			[]string{"branch", "revision", "version"},
 		),
 		proxyMetricCache: &RouteMetric{},
+		deprecatedProxyTotalGauge: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: DeprecatedHTTPProxyTotalGauge,
+				Help: "(Deprecated): Total number of HTTPProxies that exist regardless of status.",
```
Comment (skriss) [inline internal/metrics/metrics.go:103]: It'd be helpful to include the name of the metric that replaces this one in the help text -- "Use contour_httpproxy instead." at the end of the help text should be sufficient. Same goes for other metrics.

File: internal/metrics/metrics.go [hunk 5]
```diff
@@ -58,11 +64,17 @@ type Meta struct {
...
+	DeprecatedHTTPProxyRootTotalGauge = "contour_httpproxy_root_total"
+	DeprecatedHTTPProxyInvalidGauge   = "contour_httpproxy_invalid_total"
+	DeprecatedHTTPProxyValidGauge     = "contour_httpproxy_valid_total"
+	DeprecatedHTTPProxyOrphanedGauge  = "contour_httpproxy_orphaned_total"
+
+	HTTPProxyTotalGauge     = "contour_httpproxy"
+	HTTPProxyRootTotalGauge = "contour_httpproxy_root"
+	HTTPProxyInvalidGauge   = "contour_httpproxy_invalid"
+	HTTPProxyValidGauge     = "contour_httpproxy_valid"
+	HTTPProxyOrphanedGauge  = "contour_httpproxy_orphaned"
```
Comment (skriss) [inline internal/metrics/metrics.go:77]: We should update the [example grafana dashboard](https://github.com/projectcontour/contour/blob/main/examples/grafana/02-grafana-configmap.yaml) to use the new metric names.

**Review by skriss — CHANGES_REQUESTED** [review]
thanks for the PR @abhide! Couple comments on this one.

File: internal/metrics/metrics.go (hunk 5 above)
Comment (abhide) [reply internal/metrics/metrics.go:78]: Thanks. Fixed

File: internal/metrics/metrics.go (hunk 4 above)
Comment (abhide) [reply internal/metrics/metrics.go]: yes.. good suggestion. fixed

File: internal/metrics/metrics.go [hunk 6]
```diff
@@ -85,6 +98,51 @@ func NewMetrics(registry *prometheus.Registry) *Metrics {
...
+					"(Deprecated): Total number of root HTTPProxies. Note there will only be a single root HTTPProxy per vhost. Use %s instead",
+					HTTPProxyRootTotalGauge),
+			},
+			[]string{"namespace"},
+		),
+		deprecatedProxyInvalidGauge: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: DeprecatedHTTPProxyInvalidGauge,
+				Help: fmt.Sprintf(
+					"(Deprecated): Total number of invalid HTTPProxies Use %s instead.",
```
Comment (skriss) [inline internal/metrics/metrics.go:123]:
```suggestion
					"(Deprecated): Total number of invalid HTTPProxies. Use %s instead.",
```

**Review by skriss — CHANGES_REQUESTED** [review]
just one nit, then LGTM

File: internal/metrics/metrics.go (hunk 6 above)
Comment (abhide) [reply internal/metrics/metrics.go]: Fixed.

Comment (skriss) [reply internal/metrics/metrics.go]: need to update the generated output too

Comment (abhide) [reply internal/metrics/metrics.go]: Sorry.. fixed.

**Review by skriss — APPROVED** [review]
LGTM, thanks @abhide!

**Comment by abhide** [comment]
Release Note:

Following metrics have been deprecated and new metrics are added. This change is done to make the metric name follow promlint convention
The following Prometheus Gauges are renamed:
```
    contour_httpproxy_total -> contour_httpproxy
    contour_httpproxy_invalid_total  -> contour_httpproxy_invalid
    contour_httpproxy_orphaned_total  -> contour_httpproxy_orphaned
    contour_httpproxy_valid_total  -> contour_httpproxy_valid
    contour_httpproxy_root_total  -> contour_httpproxy_root
``` 
We encourage operators to have dashboard and alert queries refer to new names.
The old metrics will be removed completely in next release.
